1.Explain the key featres of Python that make it popular choice for programming language?


1.Simple and Readable Syntax:
python code is clean and easy to understand ,which makes it a great language for beginners and increases prodctivity for experienced developers

2.Interpreted language:
Python runs code line by line using an interpreter,which makes debugging easier and faster since errors can be caught as they happen

3.Object oriented andFunctional:
Python supports multiple programming,include object -oriented,functional,procedural stlyes
Python is sed in different fields like Web Development,Data science,M.l,A.I,Automation ,scripting,Finance,gaming.

2.Describe the role of predefined keywords in python and provideexamples of how they are used in a program

  control flow: Keywords like if,else,elif,for,while break,continu.e and return are used to control the order in which statements are executed in a program.They allow for decision making and repetition.
  

In [1]:
a = 10
if a > 7:
  print("a is greater than 7")
else:
  print("a is 7 or less")

a is greater than 7


2.Function and class Definitions
1.def- to define a function
2.class- to define a class
3.retrn- to return a value from function

In [2]:
def greet(name):
  return "hello," + name

print(greet("sahana"))



hello,sahana


3.Exception Handling

Try except finally raise are some keywords in this

In [14]:
try:
  result = 10/0
except ZeroDivisionError:
  print("you can't divide by zero.")

you can't divide by zero.


4.variable scope and Declaration
Global,nonlocal

In [17]:

def outer():
  a = "outer"
  def inner():
      nonlocal a
      a = "inner"
  inner()
  print(a)



5.Boolean and null values
True false none


In [18]:
True -False

1

In [19]:
f = None
type(f)

NoneType

3.Compare and constrast mutable and immutableobjects in python with examples?
Mutable Objects:
can be changed after creation
changing the object doesn't change its identity(memory address)
useful when you want to update or modify data without creating a new object.

In [20]:
list_cont = [1,2,4.3,2+5j,4.3,True,"suudha"]
list_cont

[1, 2, 4.3, (2+5j), 4.3, True, 'suudha']

In [21]:
list_cont[6] ="arun"
list_cont




















[1, 2, 4.3, (2+5j), 4.3, True, 'arun']

Immutable Objects:
1.cannot be changed after creation
2.Any operation that seems to change the object actually creates a new object
3.useful for fixed data ensuring consistency
Ex:string,float,int,tuple

In [22]:
a = "arunkumar"
a

'arunkumar'

In [23]:
a[5] = "seetha"
a

TypeError: 'str' object does not support item assignment

4. Discuss the different types of operators in Python and provide examples of how they are used?

python supports different types of operators,which are used to perform operations on variables and values.arthimetic operators,assignment operators,comparision operators,logical operators,bitwise operators,special operators.

1.Arthimetic Operators
Arithmetic operators are used to perform mathematical operations like addition, subtraction, multiplication, etc. For example,
In the above example, we have used multiple arithmetic operators,

+	Addition	5 + 2 = 7
-	Subtraction	4 - 2 = 2
*	Multiplication	2 * 3 = 6
/	Division	4 / 2 = 2
//	Floor Division	10 // 3 = 3
%	Modulo	5 % 2 = 1
**	Power	4 ** 2 = 16


In [24]:
a = 7
b = 2

# addition
print ('Sum: ', a + b)

# subtraction
print ('Subtraction: ', a - b)

# multiplication
print ('Multiplication: ', a * b)

# division
print ('Division: ', a / b)

# floor division
print ('Floor Division: ', a // b)

# modulo
print ('Modulo: ', a % b)

# a to the power b
print ('Power: ', a ** b)

Sum:  9
Subtraction:  5
Multiplication:  14
Division:  3.5
Floor Division:  3
Modulo:  1
Power:  49


2.python Assignment operators:
used to assign values to variables
Operator	Name	Example

=	Assignment Operator	a = 7
+=	Addition Assignment	a += 1 # a = a + 1
-=	Subtraction Assignment	a -= 3 # a = a - 3
*=	Multiplication Assignment	a *= 4 # a = a * 4
/=	Division Assignment	a /= 3 # a = a / 3
%=	Remainder Assignment	a %= 10 # a = a % 10
**=	Exponent Assignment	a **= 10 # a = a ** 10

In [25]:
# assign 10 to a
a = 10

# assign 5 to b
b = 5

# assign the sum of a and b to a
a += b      # a = a + b

print(a)



15


Here, we have used the += operator to assign the sum of a and b to a.

Similarly, we can use any other assignment operators as per our needs.



3.Python Comparison Operators
Comparison operators compare two values/variables and return a boolean result: True or false

Operator	Meaning	Example
==	Is Equal To	3 == 5 gives us False
!=	Not Equal To	3 != 5 gives us True
>	Greater Than	3 > 5 gives us False
<	Less Than	3 < 5 gives us True
>=	Greater Than or Equal To	3 >= 5 give us False
<=	Less Than or Equal To	3 <= 5 gives us True
For Example:

In [27]:
a = 5

b = 2

# equal to operator
print('a == b =', a == b)

# not equal to operator
print('a != b =', a != b)

# greater than operator
print('a > b =', a > b)

# less than operator
print('a < b =', a < b)

# greater than or equal to operator
print('a >= b =', a >= b)

# less than or equal to operator
print('a <= b =', a <= b)

a == b = False
a != b = True
a > b = True
a < b = False
a >= b = True
a <= b = False


4.Logical operators are used to check whether an expression is True or False. They are used in decision-making. For example,

 Operator	Example	Meaning
and	a and b	Logical AND:
True only if both the operands are True
or	a or b	Logical OR:
True if at least one of the operands is True
not	not a	Logical NOT:
True if the operand is False and vice-versa.

In [28]:
# logical AND
print(True and True)
print(True and False)

# logical OR
print(True or False)

# logical NOT
print(not True)

True
False
True
False


5.Bitwise operators act on operands as if they were strings of binary digits. They operate bit by bit,

For example, 2 is 10 in binary, and 7 is 111
In the table below: Let x = 10 (0000 1010 in binary) and y = 4 (0000 0100 in binary)

Operator	Meaning	Example
&	Bitwise AND	x & y = 0 (0000 0000)
|	Bitwise OR	x | y = 14 (0000 1110)
~	Bitwise NOT	~x = -11 (1111 0101)
^	Bitwise XOR	x ^ y = 14 (0000 1110)
>>	Bitwise right shift	x >> 2 = 2 (0000 0010)
<<	Bitwise left shift	x 0010 1000)

Operator	Meaning	Example
&	Bitwise AND	x & y = 0 (0000 0000)
|	Bitwise OR	x | y = 14 (0000 1110)
~	Bitwise NOT	~x = -11 (1111 0101)
^	Bitwise XOR	x ^ y = 14 (0000 1110)
>>	Bitwise right shift	x >> 2 = 2 (0000 0010)
<<	Bitwise left shift	x 0010 1000)

6. Python Special operators

Python language offers some special types of operators like the identity operator and the membership operator. They are described below with examples.

Identity operators
In Python, is and is not are used to check if two values are located at the same memory location.

It's important to note that having two variables with equal values doesn't necessarily mean they are identical.

Operator	 Meaning 	Example
is	 True if the operands are identical (refer to the same object)	x is True
is not	True if the operands are not identical (do not refer to the same object)	x is not True



In [29]:
x1 = 5
y1 = 5
x2 = 'Hello'
y2 = 'Hello'
x3 = [1,2,3]
y3 = [1,2,3]

print(x1 is not y1)

print(x2 is y2)

print(x3 is y3)

False
True
False


Here, we see that x1 and y1 are integers of the same values, so they are equal as well as identical. The same is the case with x2 and y2 (strings).

But x3 and y3 are lists. They are equal but not identical. It is because the interpreter locates them separately in memory, although they are equal.

Membership operators

In Python, in and not in are the membership operators. They are used to test whether a value or variable is found in a sequence (string, list, tuple, set and dictionary).


In a dictionary, we can only test for the presence of a key, not the value.

Operator 	            Meaning	                                      Example       
in	       True if value/variable is found in the sequence	     5 in x
not in	   True if value/variable is not found in the sequence	 5 not in x


In [30]:
message = 'Hello world'
dict1 = {1:'a', 2:'b'}

# check if 'H' is present in message string
print('H' in message)  # prints True

# check if 'hello' is present in message string
print('hello' not in message)  # prints True

# check if '1' key is present in dict1
print(1 in dict1)  # prints True

# check if 'a' key is present in dict1
print('a' in dict1)  # prints False


True
True
True
False


5. Explain the concept of type casting in Python with examples

The conversion of one data type into other data type is called type casting.
It is also known as type conversion.This is common when you need to perform operations involving different data types.

Python Implicit Type Conversion
In some situations, Python automatically converts one data type to another. This is known as implicit type conversion.


 Converting integer to float
Let's see an example where Python promotes the conversion of the lower data type (integer) to the higher data type (float) to avoid data loss.



In [31]:
integer_number = 123
float_number = 1.23

new_number = integer_number + float_number

# display new value and resulting data type
print("Value:",new_number)
print("Data Type:",type(new_number))

Value: 124.23
Data Type: <class 'float'>


In the above example, we have created two variables: integer_number and float_number of int and float type respectively.

Then we added these two variables and stored the result in new_number
As we can see new_number has value 124.23 and is of the float data type.

It is because Python always converts smaller data types to larger data types to avoid the loss of data.

Explicit Type Conversion
In Explicit Type Conversion, users convert the data type of an object to required data type.

We use the built-in functions like int(), float(), str(), etc to perform explicit type conversion.

This type of conversion is also called typecasting because the user casts (changes) the data type of the objects.

In [32]:
num_string = '12'
num_integer = 23

print("Data type of num_string before Type Casting:",type(num_string))

# explicit type conversion
num_string = int(num_string)

print("Data type of num_string after Type Casting:",type(num_string))

num_sum = num_integer + num_string

print("Sum:",num_sum)
print("Data type of num_sum:",type(num_sum))

Data type of num_string before Type Casting: <class 'str'>
Data type of num_string after Type Casting: <class 'int'>
Sum: 35
Data type of num_sum: <class 'int'>


In the above example, we have created two variables: num_string and num_integer with str and int type values respectively.

Here, we have used int() to perform explicit type conversion of num_string to integer type.

After converting num_string to an integer value, Python is able to add these two variables.

Finally, we got the num_sum value i.e 35 and data type to be int.

6.How do conditional statements work in python illustrate with example.

conditional statements are used to make decisions in code. based on whether a condition is true or false,different blocks of code are executed.

1.if statement
Execute the block of code if the condition is true.


In [33]:
a = 20
if a > 10:
  print("a is greater than 10")



a is greater than 10


2.it else statement
 adds an alternative block if  the  condition is false

In [35]:
b = 5
if b > 9:
  print("b is greater than 9")
else:
  print("b is 9 or less")

b is 9 or less


if elif statement
used when we have multiple condition to check


In [37]:
x = 7
if x > 12:
  print("x is greater than 12")
elif x == 7:
  print("x is equal to 7")
else:
  print("x is less than 12 and not equal to 7")

x is equal to 7


nested if
an if statement inside another if statement


In [38]:
c = 25
if c > 20:
  if c < 30:
    print("c is between 20 and 30")


c is between 20 and 30


7. Describe the different types of loops in Python and their use cases with examples.

loops in python
loops are uused to execute a block of code repeatedly.python spports mainly two types of loops

1.FOR Loop
Used to iterate over a seqence like list,tuple,string or a range of numbers
syn
for variable in sequence:
    # code block

In [39]:
colors =["white", "blue", "black"]
for color in colors:
  print(color)

white
blue
black


use case
iterating through items in a list or string.
performing a fixed number of repetitions using range()

In [40]:
# print numbers 1 to 5
for i in range(1, 7):
  print(i)

1
2
3
4
5
6


2.while loop
Repeats as long as acondition is true

Syntax
while conditiom:
    # code block

In [41]:
s = 2
while s <= 6:
  print(s)
  s += 2



2
4
6


use cases repeating tasks  until a certain condition changes waiting for user input or a status  to change


3. Nested loops
loops are formed inside of the loops

In [45]:
for i in range(1, 4):
  for j in range(1, 4):
       print(f"i = {i}", f"j = {j}")

i = 1 j = 1
i = 1 j = 2
i = 1 j = 3
i = 2 j = 1
i = 2 j = 2
i = 2 j = 3
i = 3 j = 1
i = 3 j = 2
i = 3 j = 3


loop control statements
break - exists the loop
continue skips to the next iteration
pass does nothing (uses as place holder)

In [46]:
for i in range(1, 10):
    if i == 5:
      break
    print(i)

1
2
3
4
